data/trie/extensionNode.go

```diff
@@ -285,63 +325,68 @@ func (en *extensionNode) getNext(key []byte, db data.DBWriteCacher) (node, []byt
 	return en.child, key, nil
 }
 
-func (en *extensionNode) insert(n *leafNode, db data.DBWriteCacher) (bool, node, [][]byte, error) {
+func (en *extensionNode) insert(n *leafNode, db data.DBWriteCacher) (node, [][]byte, error) {
 	emptyHashes := make([][]byte, 0)
 	err := en.isEmptyOrNil()
 	if err != nil {
-		return false, nil, emptyHashes, fmt.Errorf("insert error %w", err)
+		return nil, emptyHashes, fmt.Errorf("insert error %w", err)
 	}
 	err = resolveIfCollapsed(en, 0, db)
 	if err != nil {
-		return false, nil, emptyHashes, err
+		return nil, emptyHashes, err
 	}
+
 	keyMatchLen := prefixLen(n.Key, en.Key)
 
 	// If the whole key matches, keep this extension node as is
 	// and only update the value.
 	if keyMatchLen == len(en.Key) {
-		var dirty bool
-		var newNode, newEn node
-		var oldHashes [][]byte
-
-		n.Key = n.Key[keyMatchLen:]
-		dirty, newNode, oldHashes, err = en.child.insert(n, db)
-		if !dirty || err != nil {
-			return false, en, emptyHashes, err
-		}
+		return en.insertInSameEn(n, keyMatchLen, db)
+	}
 
-		if !en.dirty {
-			oldHashes = append(oldHashes, en.hash)
-		}
+	// Otherwise branch out at the index where they differ.
+	return en.insertInNewBn(n, keyMatchLen)
+}
 
-		newEn, err = newExtensionNode(en.Key, newNode, en.marsh, en.hasher)
-		if err != nil {
-			return false, nil, emptyHashes, err
-		}
+func (en *extensionNode) insertInSameEn(n *leafNode, keyMatchLen int, db data.DBWriteCacher) (node, [][]byte, error) {
+	n.Key = n.Key[keyMatchLen:]
+	newNode, oldHashes, err := en.child.insert(n, db)
+	if check.IfNil(newNode) || err != nil {
+		return nil, [][]byte{}, err
```

Comment:
Could you have nil newNode and nil err? 
then you end up with nil err

if it cannot happen then you don't need the check.IfNil right?

Reply:
There is a case where you can have a nil newNode and a nil err: if you try to insert the same value at the same key (`leafNode.go` L233). It is not an error, but there is no newNode to return. The `insert()` return is treated in `particiaMerkleTrie.go` on L132 - L138.